File: end-of-line-trim/typescript/src/endOfLineTrim.ts
export function trim(input: string): string {
  let result = '';
  let lineStart = 0;
  let i = 0;
  while (i < input.length) {
    const ch = input[i];
    if (ch === '\n') {
      result += rightTrim(input, lineStart, i) + '\n';
      lineStart = i + 1;
      i++;
    } else if (ch === '\r' && input[i + 1] === '\n') {
      result += rightTrim(input, lineStart, i) + '\r\n';
      lineStart = i + 2;
      i += 2;
    } else {
      i++;
    }
  }
  result += rightTrim(input, lineStart, input.length);
  return result;
}

function rightTrim(input: string, start: number, end: number): string {
  let trimmedEnd = end;
  while (trimmedEnd > start) {
    const c = input[trimmedEnd - 1];
    if (c !== ' ' && c !== '\t') break;
    trimmedEnd--;
  }
  return input.slice(start, trimmedEnd);
}

File: end-of-line-trim/python/src/end_of_line_trim/end_of_line_trim.py
from __future__ import annotations


def trim(input: str) -> str:
    parts: list[str] = []
    line_start = 0
    i = 0
    n = len(input)
    while i < n:
        ch = input[i]
        if ch == "\n":
            parts.append(_right_trim(input, line_start, i))
            parts.append("\n")
            line_start = i + 1
            i += 1
        elif ch == "\r" and i + 1 < n and input[i + 1] == "\n":
            parts.append(_right_trim(input, line_start, i))
            parts.append("\r\n")
            line_start = i + 2
            i += 2
        else:
            i += 1
    parts.append(_right_trim(input, line_start, n))
    return "".join(parts)


def _right_trim(input: str, start: int, end: int) -> str:
    trimmed_end = end
    while trimmed_end > start and input[trimmed_end - 1] in (" ", "\t"):
        trimmed_end -= 1
    return input[start:trimmed_end]
